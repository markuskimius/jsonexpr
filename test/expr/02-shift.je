PRINT("== OPERATION ==");
PRINT(0xF0 << 4);
PRINT(0xF0 >> 4);
PRINT(0xF0 >>> 4);
PRINT();

PRINT(~0xF0 << 4);
PRINT(~0xF0 >> 4);
PRINT(~0xF0 >>> 4);
PRINT();

PRINT("== NOSHIFT ==");
PRINT(-1 << 0);
PRINT(-1 >> 0);
PRINT(-1 >>> 0);
PRINT();

PRINT("== ONESHIFT ==");
PRINT(-1 << 1);
PRINT(-1 >> 1);
PRINT(-1 >>> 1);
PRINT();

PRINT(-1 << -1);
PRINT(-1 >> -1);
PRINT(-1 >>> -1);
PRINT();

PRINT("== OVERFLOW ==");
PRINT(-1 << 63);
PRINT(-1 << 64);
PRINT(-1 << 65);
PRINT();

PRINT(-1 >> 63);
PRINT(-1 >> 64);
PRINT(-1 >> 65);
PRINT();

PRINT(-1 >>> 63);
PRINT(-1 >>> 64);
PRINT(-1 >>> 65);
PRINT();

PRINT(0 << 63);
PRINT(0 << 64);
PRINT(0 << 65);
PRINT();

PRINT(0 >> 63);
PRINT(0 >> 64);
PRINT(0 >> 65);
PRINT();

PRINT(0 >>> 63);
PRINT(0 >>> 64);
PRINT(0 >>> 65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF << 62);
PRINT(0x7FFFFFFFFFFFFFFF << 63);
PRINT(0x7FFFFFFFFFFFFFFF << 64);
PRINT(0x7FFFFFFFFFFFFFFF << 65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF >> 62);
PRINT(0x7FFFFFFFFFFFFFFF >> 63);
PRINT(0x7FFFFFFFFFFFFFFF >> 64);
PRINT(0x7FFFFFFFFFFFFFFF >> 65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF >>> 62);
PRINT(0x7FFFFFFFFFFFFFFF >>> 63);
PRINT(0x7FFFFFFFFFFFFFFF >>> 64);
PRINT(0x7FFFFFFFFFFFFFFF >>> 65);
PRINT();

PRINT("== UNDERFLOW ==");
PRINT(-1 << -63);
PRINT(-1 << -64);
PRINT(-1 << -65);
PRINT();

PRINT(-1 >> -63);
PRINT(-1 >> -64);
PRINT(-1 >> -65);
PRINT();

PRINT(-1 >>> -63);
PRINT(-1 >>> -64);
PRINT(-1 >>> -65);
PRINT();

PRINT(0 << -63);
PRINT(0 << -64);
PRINT(0 << -65);
PRINT();

PRINT(0 >> -63);
PRINT(0 >> -64);
PRINT(0 >> -65);
PRINT();

PRINT(0 >>> -63);
PRINT(0 >>> -64);
PRINT(0 >>> -65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF << -62);
PRINT(0x7FFFFFFFFFFFFFFF << -63);
PRINT(0x7FFFFFFFFFFFFFFF << -64);
PRINT(0x7FFFFFFFFFFFFFFF << -65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF >> -62);
PRINT(0x7FFFFFFFFFFFFFFF >> -63);
PRINT(0x7FFFFFFFFFFFFFFF >> -64);
PRINT(0x7FFFFFFFFFFFFFFF >> -65);
PRINT();

PRINT(0x7FFFFFFFFFFFFFFF >>> -62);
PRINT(0x7FFFFFFFFFFFFFFF >>> -63);
PRINT(0x7FFFFFFFFFFFFFFF >>> -64);
PRINT(0x7FFFFFFFFFFFFFFF >>> -65);
PRINT();

PRINT("== PRECEDENCE ==");
PRINT(~0xF0 << 4);
PRINT(0xFF00 & 0xF0F0 << 4);
PRINT(0xFF00 << 4 & 0xF0F0);
PRINT();

PRINT(0xF00 << 8 >> 4);
PRINT(0xF00 >> 8 << 4);
PRINT(0xF00 >>> 8 << 4);
PRINT();

PRINT(~0xF00 << 8 >> 4);
PRINT(~0xF00 >> 8 << 4);
PRINT(~0xF00 >>> 8 << 4);
PRINT();
