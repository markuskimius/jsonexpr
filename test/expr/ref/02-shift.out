== OPERATION ==
3840
15
15

-3856
-16
1152921504606846960

== NOSHIFT ==
-1
-1
-1

== ONESHIFT ==
-2
-1
9223372036854775807

-1
-2
-2

== OVERFLOW ==
-9223372036854775808
0
0

-1
-1
-1

1
0
0

0
0
0

0
0
0

0
0
0

-4611686018427387904
-9223372036854775808
0
0

1
0
0
0

1
0
0
0

== UNDERFLOW ==
-1
-1
-1

-9223372036854775808
0
0

-9223372036854775808
0
0

0
0
0

0
0
0

0
0
0

1
0
0
0

-4611686018427387904
-9223372036854775808
0
0

-4611686018427387904
-9223372036854775808
0
0

== PRECEDENCE ==
-3856
3840
61440

61440
240
240

-61456
-256
1152921504606846720

